File: chapters/09_multi_agent_routing/v2_central_router.ts
// TypeScript mirror for Chapter 09: Multi-Agent Routing (v2)
// Run with: npx ts-node v2_central_router.ts

const BINDINGS: Record<string, string> = {
  "build ui": "frontend_agent",
  "implement api": "backend_agent",
  "write test": "qa_agent",
};

function routeTask(task: string, bindings?: Record<string, string>): string {
  return routeTaskWithMatch(task, bindings).agent;
}

function routeTaskWithMatch(
  task: string,
  bindings?: Record<string, string>
): { agent: string; matched_by: string } {
  const useBindings = bindings || BINDINGS;
  for (const [keyword, agent] of Object.entries(useBindings)) {
    if (task.includes(keyword)) {
      return { agent, matched_by: keyword };
    }
  }
  return { agent: "generalist_agent", matched_by: "default" };
}

function runCentralized(tasks: string[]): string[] {
  const logs: string[] = [];
  for (const task of tasks) {
    const agent = routeTask(task);
    logs.push(`routed:${task}->${agent}`);
  }
  return logs;
}

function main(): void {
  console.log(runCentralized(["build ui", "implement api", "unknown"]));
}

main();

File: chapters/09_multi_agent_routing/v2_central_router.py
from __future__ import annotations


BINDINGS = {
    "build ui": "frontend_agent",
    "implement api": "backend_agent",
    "write test": "qa_agent",
}


def route_task(task: str, bindings: dict[str, str] | None = None) -> str:
    return route_task_with_match(task, bindings)["agent"]


def route_task_with_match(task: str, bindings: dict[str, str] | None = None) -> dict[str, str]:
    use_bindings = bindings or BINDINGS
    for keyword, agent in use_bindings.items():
        if keyword in task:
            return {
                "agent": agent,
                "matched_by": keyword,
            }
    return {
        "agent": "generalist_agent",
        "matched_by": "default",
    }


def run_centralized(tasks: list[str]) -> list[str]:
    logs: list[str] = []
    for task in tasks:
        agent = route_task(task)
        logs.append(f"routed:{task}->{agent}")
    return logs


def main() -> None:
    print(run_centralized(["build ui", "implement api", "unknown"] ))


if __name__ == "__main__":
    main()
